File: src/downloader/logger.ts
import { formatWithOptions } from "node:util";

export type LogFields = Record<string, unknown>;

export type Logger = Record<
  LogLevel,
  (event: string, message?: string, fields?: LogFields) => void
> & {
  configure: (options: CreateLoggerOptions) => void;
  log: (
    event: string,
    message?: string,
    fieldsByLevel?: Partial<Record<LogLevel, LogFields>>,
  ) => void;
  raw: (line: string) => void;
};

export type LogLevel = "debug" | "error" | "info" | "silent" | "trace" | "warn";

interface CreateLoggerOptions {
  format?: "json" | "pretty";
  level?: LogLevel;
  write?: (line: string) => void;
}

const LEVEL_WEIGHTS: Record<LogLevel, number> = {
  debug: 10,
  error: 40,
  info: 20,
  silent: 0,
  trace: 5,
  warn: 30,
};

function createLogger({
  format = "json",
  level: minimumLevel = "info",
  write = (line: string) => process.stdout.write(`${line}\n`),
}: CreateLoggerOptions = {}): Logger {
  const createLogFn = (level: LogLevel) =>
    function log(event: string, message?: string, fields: LogFields = {}) {
      const levelWeight = LEVEL_WEIGHTS[level];
      if (!levelWeight || levelWeight < LEVEL_WEIGHTS[minimumLevel]) {
        return;
      }

      const payload = {
        event,
        level,
        msg: message,
        time: new Date().toISOString(),
        ...fields,
      };
      write(
        format === "json"
          ? JSON.stringify(payload)
          : `${payload.time} ${level.toUpperCase()} ${event} ${message?.concat(" ")}${formatWithOptions({ colors: true, depth: Infinity }, fields)}`,
      );
    };

  return {
    configure({
      format: newFormat = format,
      level: newMinimumLevel = minimumLevel,
      write: newWrite = write,
    }: CreateLoggerOptions) {
      format = newFormat;
      minimumLevel = newMinimumLevel;
      write = newWrite;
    },
    debug: createLogFn("debug"),
    error: createLogFn("error"),
    info: createLogFn("info"),
    log: (event, message, fieldsByLevel) => {
      const levels: LogLevel[] = ["trace", "debug", "info", "warn", "error"];
      for (const level of levels) {
        const fields = fieldsByLevel?.[level];
        if (!fields) continue;
        createLogFn(level)(event, message, fields);
        return;
      }
    },
    raw: (line: string) => {
      write(line);
    },
    silent: createLogFn("silent"),
    trace: createLogFn("trace"),
    warn: createLogFn("warn"),
  };
}

export const logger = createLogger();

File: src/downloader/logger.rs
use std::collections::BTreeMap;
use std::sync::Arc;

use serde_json::{Map, Value, json};

#[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd)]
pub enum LogLevel {
    Debug,
    Info,
    Warn,
    Error,
}

impl LogLevel {
    #[must_use]
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Debug => "debug",
            Self::Info => "info",
            Self::Warn => "warn",
            Self::Error => "error",
        }
    }
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum LogFormat {
    Json,
    Pretty,
}

pub type LogFields = Map<String, Value>;
pub type LogWriter = Arc<dyn Fn(String) + Send + Sync>;

#[derive(Clone)]
pub struct Logger {
    format: LogFormat,
    minimum_level: LogLevel,
    writer: Option<LogWriter>,
}

impl Logger {
    #[must_use]
    pub fn new(format: LogFormat, minimum_level: LogLevel, writer: LogWriter) -> Self {
        Self {
            format,
            minimum_level,
            writer: Some(writer),
        }
    }

    #[must_use]
    pub fn silent() -> Self {
        Self {
            format: LogFormat::Json,
            minimum_level: LogLevel::Error,
            writer: None,
        }
    }

    pub fn debug(&self, event: &str, message: &str, fields: LogFields) {
        self.log(LogLevel::Debug, event, message, fields);
    }

    pub fn info(&self, event: &str, message: &str, fields: LogFields) {
        self.log(LogLevel::Info, event, message, fields);
    }

    pub fn warn(&self, event: &str, message: &str, fields: LogFields) {
        self.log(LogLevel::Warn, event, message, fields);
    }

    pub fn error(&self, event: &str, message: &str, fields: LogFields) {
        self.log(LogLevel::Error, event, message, fields);
    }

    fn log(&self, level: LogLevel, event: &str, message: &str, fields: LogFields) {
        if level < self.minimum_level {
            return;
        }
        let Some(writer) = &self.writer else {
            return;
        };
        let time = chrono::Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Millis, true);
        let line = match self.format {
            LogFormat::Json => {
                let mut entry = fields;
                entry.insert("event".into(), json!(event));
                entry.insert("level".into(), json!(level.as_str()));
                entry.insert("msg".into(), json!(message));
                entry.insert("time".into(), json!(time));
                Value::Object(entry).to_string()
            }
            LogFormat::Pretty => {
                let ordered = BTreeMap::from_iter(fields.into_iter());
                format!("{time} {} {event} {message} {}", level.as_str().to_uppercase(), serde_json::to_string(&ordered).unwrap())
            }
        };
        writer(line);
    }
}
